tf_quant_finance/models/heston/approximations/asian_prices.py

```diff
@@ -0,0 +1,509 @@
+      dividend_rates=tf.cast(dividend_rates, np.complex128),
+      variances=tf.cast(variances, np.complex128),
+      mean_reversion=tf.cast(mean_reversion, np.complex128),
+      volvol=tf.cast(volvol, np.complex128),
+      rho=tf.cast(rho, np.complex128),
+      theta=tf.cast(theta, np.complex128),
+      strikes=tf.cast(strikes, np.complex128),
+      start_times=tf.cast(start_times, np.complex128),
+      expiries=tf.cast(expiries, np.complex128),
+      sampling_times=tf.cast(sampling_times, np.complex128),
+      k_tensor=tf.cast(k_tensor, np.complex128),
+      d_tau_k=tf.cast(d_tau_k, np.complex128),
+      k_star=k_star,
+      n=n,
+      dtype=np.complex128
+    )
+
+    if integration_method is None:
+      integration_method = integration.IntegrationMethod.GAUSS_LEGENDRE
+
+    f = lambda x: tf.cast(asian_prices_handler.integrand(x), dtype)
+
+    if 'lower' not in kwargs:
+      kwargs['lower'] = 1e-9
+    if 'upper' not in kwargs:
+      kwargs['upper'] = 100
+
+    kwargs['lower'] = tf.constant(kwargs['lower'], dtype=dtype)
+    kwargs['upper'] = tf.constant(kwargs['upper'], dtype=dtype)
+
+    integral = integration.integrate(
+      f,
+      method=integration_method,
+      dtype=dtype,
+      **kwargs
+    )
+
+    asian_forward = tf.cast(
+      tf.math.real(asian_prices_handler.Phi(1, 0)),
+      dtype=dtype
+    )
+    dcfs = tf.math.exp(-discount_rates * expiries)
+
+    calls = prefactor * dcfs * ((asian_forward - strikes) / 2
+                                + tf.expand_dims(integral, -1) / np.pi)
+
+    if is_call_options is None:
+      return tf.reshape(calls, results_shape)
+
+    puts = prefactor * dcfs * ((strikes - asian_forward) / 2
+                               + tf.expand_dims(integral, -1) / np.pi)
+    predicate = tf.broadcast_to(is_call_options, tf.shape(calls))
+    return tf.reshape(tf.where(predicate, calls, puts), results_shape)
+
+
+@utils.dataclass
+class _AsianPricesHandler:
+  """Handles the various pricing functions for asian options."""
+  spots: types.RealTensor
+  discount_rates: types.RealTensor
+  dividend_rates: types.RealTensor
+  variances: types.RealTensor
+  mean_reversion: types.RealTensor
+  theta: types.RealTensor
+  volvol: types.RealTensor
+  rho: types.RealTensor
+  start_times: types.RealTensor
+  expiries: types.RealTensor
+  strikes: types.RealTensor
+  sampling_times: types.RealTensor
+  k_tensor: types.RealTensor
+  d_tau_k: types.RealTensor
+  k_star: int
+  n: int
+  dtype: tf.DType
+
+  def F(self, z1, z2, tau):
+    """Calculates F(z1, z2) (eq. (11) from the referenced paper)"""
+    x = tf.math.sqrt(tf.math.pow(self.mean_reversion, 2)
+                     - 2 * z1 * tf.math.pow(self.volvol, 2))
+
+    term_1 = tf.math.cosh(0.5 * tau * x)
+    term_2 = ((self.mean_reversion - z2 * tf.math.pow(self.volvol, 2))
+               * tf.math.sinh(0.5 * tau * x)) / x
+
+    return term_1 + term_2
+
+  def F_tilde(self, z1, z2, tau):
```

Comment:
nit: f_tilde to follow Python style guide

Reply:
As above - this is to match two equations in the reference paper, but happy to defer if you feel very strongly here.